// @nguniversal/common/esm2015/clover/src/styles_host.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { DOCUMENT, ɵgetDOM as getDOM } from '@angular/common';
import { APP_ID, Inject, Injectable, Optional } from '@angular/core';
import { ɵSharedStylesHost as SharedStylesHost } from '@angular/platform-browser';
export class SSRStylesHost extends SharedStylesHost {
    constructor(doc, appId) {
        var _a;
        super();
        this.doc = doc;
        this.appId = appId;
        this._styleNodes = new Set();
        this.head = this.doc.querySelector('head');
        const styles = (_a = this.head) === null || _a === void 0 ? void 0 : _a.querySelectorAll(`style[ng-style='${this.appId}']`);
        if (styles === null || styles === void 0 ? void 0 : styles.length) {
            const items = Array.from(styles);
            this._styleNodesInDOM = new Map(items.map((el) => [el.textContent, el]));
        }
    }
    _addStyle(style) {
        var _a, _b;
        const element = (_a = this._styleNodesInDOM) === null || _a === void 0 ? void 0 : _a.get(style);
        if (element) {
            if (typeof ngDevMode !== undefined && ngDevMode) {
                element.setAttribute('_ng-style-re-used', '');
            }
            (_b = this._styleNodesInDOM) === null || _b === void 0 ? void 0 : _b.delete(style);
            this._styleNodes.add(element);
            return;
        }
        const el = getDOM().createElement('style');
        el.textContent = style;
        if (this.appId) {
            el.setAttribute('ng-style', this.appId);
        }
        if (this.head) {
            this.head.appendChild(el);
        }
        this._styleNodes.add(el);
    }
    onStylesAdded(additions) {
        additions.forEach((style) => this._addStyle(style));
    }
    addHost(_hostNode) {
        // stub
    }
    removeHost(_hostNode) {
        // stub
    }
    ngOnDestroy() {
        this._styleNodes.forEach((styleNode) => styleNode.remove());
    }
}
SSRStylesHost.decorators = [
    { type: Injectable }
];
SSRStylesHost.ctorParameters = () => [
    { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: String, decorators: [{ type: Optional }, { type: Inject, args: [APP_ID,] }] }
];
