// @nguniversal/common/clover/server/src/inline-css-processor.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@nguniversal/common/clover/server/src/inline-css-processor", ["require", "exports", "tslib", "critters", "fs"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.InlineCriticalCssProcessor = void 0;
    const tslib_1 = require("tslib");
    const critters_1 = tslib_1.__importDefault(require("critters"));
    const fs_1 = require("fs");
    class CrittersExtended extends critters_1.default {
        constructor(optionsExtended, resourceCache) {
            super({
                logger: {
                    warn: (s) => this.warnings.push(s),
                    error: (s) => this.errors.push(s),
                    info: () => { },
                },
                logLevel: 'warn',
                path: optionsExtended.outputPath,
                publicPath: optionsExtended.deployUrl,
                compress: !!optionsExtended.minify,
                pruneSource: false,
                reduceInlineStyles: false,
                mergeStylesheets: false,
                preload: 'media',
                noscriptFallback: true,
                inlineFonts: true,
            });
            this.optionsExtended = optionsExtended;
            this.resourceCache = resourceCache;
            this.warnings = [];
            this.errors = [];
        }
        readFile(path) {
            return tslib_1.__awaiter(this, void 0, void 0, function* () {
                let resourceContent = this.resourceCache.get(path);
                if (resourceContent === undefined) {
                    resourceContent = yield fs_1.promises.readFile(path);
                    this.resourceCache.set(path, resourceContent);
                }
                return resourceContent.toString();
            });
        }
    }
    class InlineCriticalCssProcessor {
        constructor(options, resourceCache) {
            this.options = options;
            this.resourceCache = resourceCache;
        }
        process(html, options) {
            return tslib_1.__awaiter(this, void 0, void 0, function* () {
                const critters = new CrittersExtended(Object.assign(Object.assign({}, this.options), options), this.resourceCache);
                const content = yield critters.process(html);
                return {
                    content,
                    errors: critters.errors.length ? critters.errors : undefined,
                    warnings: critters.warnings.length ? critters.warnings : undefined,
                };
            });
        }
    }
    exports.InlineCriticalCssProcessor = InlineCriticalCssProcessor;
});
